File: 2344-escape-the-spreading-fire/escape-the-spreading-fire.cpp
class Solution {
public:
    int n, m;
    vector<vector<int>> vis;
    pair<int, int> directions[4] = {{-1, 0}, {0, 1}, {1, 0}, {0, -1}};
    bool canReach(int waitTime, vector<vector<int>>& grid, vector<vector<int>>& fireTime)
    {
        queue<pair<int, int>> q;
        q.push({0, 0});

        // Reset visitation matrix
        vis.assign(n, vector<int>(m, 0)); 
        vis[0][0] = 1;

        int time = waitTime;
        while (!q.empty()) 
        {
            int levelSize = q.size();
            while(levelSize--)
            {
                auto [row, col] = q.front();
                q.pop();

                for (auto [dx, dy] : directions) 
                {
                    int nRow = row + dx;
                    int nCol = col + dy;

                    if(isValidCell(nRow, nCol) && grid[nRow][nCol] == 0 && !vis[nRow][nCol]) 
                    {
                        if(nRow == n - 1 && nCol == m - 1)
                        {
                            // if(fireTime[n - 1][m - 1] == 1e9) // Fire never reaches exit
                            // return true;

                            return time + 1 <= fireTime[nRow][nCol];
                        }

                        if(time + 1 < fireTime[nRow][nCol]) 
                        {
                            vis[nRow][nCol] = 1;
                            q.push({nRow, nCol});
                        }
                    }
                }
            }
            time++;
        }
        return false;
    }

    vector<vector<int>> fireSpread(vector<vector<int>>& grid) 
    {
        // 1e9 means fire never reaches
        vector<vector<int>> fireTime(n, vector<int>(m, 1e9)); 

        queue<pair<int, int>> q;
        // Multi-source BFS: Add all fire positions (1) to the queue
        for (int i = 0; i < n; i++) 
        {
            for (int j = 0; j < m; j++) 
            {
                if (grid[i][j] == 1) 
                {
                    q.push({i, j});
                    fireTime[i][j] = 0; // Fire starts here
                }
            }
        }

        // BFS
        while (!q.empty()) 
        {
            auto [row, col] = q.front();
            q.pop();

            for (auto [dx, dy] : directions) 
            {
                int nRow = row + dx;
                int nCol = col + dy;

                if (isValidCell(nRow, nCol) && grid[nRow][nCol] == 0 && fireTime[nRow][nCol] == 1e9) 
                {
                    fireTime[nRow][nCol] = fireTime[row][col] + 1;
                    q.push({nRow, nCol});
                }
            }
        }

        return fireTime;
    }

    int maximumMinutes(vector<vector<int>>& grid) 
    {
        n = grid.size();
        m = grid[0].size();

        // build the 2D time array 
        auto fireTime = fireSpread(grid);

        int start = 0;
        int end = 1e9;
        int minStayTime = -1;
        while(start <= end)
        {
            int mid = start + (end - start) / 2;
            if(canReach(mid, grid, fireTime))
            {
                minStayTime = mid;
                start = mid + 1;
            }

            else
            end = mid - 1;
        }

        // If fire never reaches exit, return max possible time
        if(minStayTime != -1 && fireTime[n-1][m-1] == 1e9)
        return 1e9;

        return minStayTime;
    }

    bool isValidCell(int row, int col)
    {
       return row >= 0 && row < n && col >= 0 && col < m;
    }
};

File: 2344-escape-the-spreading-fire/escape-the-spreading-fire.java
class Solution {

    private int n, m;
    private final int[][] directions = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
    private final int INF = 1000000000;
    private boolean[][] visited;
    private boolean canFireReachSafeHouse(int waitingTime, int[][] grid, int[][] fireTime)
    {
        // reassign the vis array for reuse
        for(int i = 0; i < n; i++){
            Arrays.fill(visited[i], false);
        }

        Queue<int[]> queue = new LinkedList<>();
        queue.offer(new int[] {0, 0});
        visited[0][0] = true;

        int time = waitingTime;
        while (!queue.isEmpty()) {
            int level = queue.size();
            for (int i = 0; i < level; i++) {
                var currentCell = queue.poll();
                int row = currentCell[0];
                int col = currentCell[1];

                for(int[] direction : directions) {
                    int nextRow = currentCell[0] + direction[0];
                    int nextCol = currentCell[1] + direction[1];
                    if(isValidCell(nextRow, nextCol) && grid[nextRow][nextCol] == 0 && !visited[nextRow][nextCol]) {
                        if(nextRow == n - 1 && nextCol == m - 1) {
                            return time + 1 <= fireTime[nextRow][nextCol];
                        }
                        if(time + 1 < fireTime[nextRow][nextCol]) {
                           visited[nextRow][nextCol] = true;
                           queue.offer(new int[] {nextRow, nextCol});
                        }
                    }
                }
            }
            time++;
        }
        return false;
    }

    private void findFireTime(int[][] grid, int[][] fireTime)
    {
        Queue<int[]> queue = new LinkedList<>();
        for(int i = 0; i < grid.length; i++) {
            for(int j = 0; j < grid[0].length; j++) {
                // add the fire cells in the queue
                if(grid[i][j] == 1) {
                    fireTime[i][j] = 0;
                    queue.offer(new int[] {i, j});
                }
                // wall and grass
                else {
                    fireTime[i][j] = INF;
                }
            }
        }

        while(!queue.isEmpty()) {
                var cell = queue.poll();
                int row = cell[0];
                int col = cell[1];

                for(var direction : directions){
                    int nextRow = row + direction[0];
                    int nextCol = col + direction[1];
                    if(isValidCell(nextRow, nextCol) && grid[nextRow][nextCol] == 0 && fireTime[nextRow][nextCol] == INF){
                        queue.offer(new int[] {nextRow, nextCol});
                        // Mark as visited now to avoid adding it multiple fireTimes
                        fireTime[nextRow][nextCol] = fireTime[row][col] + 1;
                    }
                }
        }
    }

    private boolean isValidCell(int row, int col)
    {
        return row >= 0 && row < n && col >= 0 && col < m;
    }

    private int binarySearch(int[][] grid, int[][] fireTime){
        int start = 0;
        int end = INF;
        int result = -1;
        while(start <= end){
            int mid = start + (end - start)/2;
            if(canFireReachSafeHouse(mid, grid, fireTime)){
                result = mid;
                start = mid + 1;
            }
            else{
                end = mid - 1;
            }
        }
        return result;
    }
    public int maximumMinutes(int[][] grid)
    {
        n = grid.length;
        m = grid[0].length;
        visited = new boolean[n][m];
        int[][] fireTime = new int[n][m];
        findFireTime(grid, fireTime);

        int minStayTime = binarySearch(grid, fireTime);
        // If fire never reaches exit but the person can reach exit, return max possible time
        if(minStayTime != -1 && fireTime[n - 1][m - 1] == INF) {
            return INF;
        }
        return minStayTime;
    }
}